internal/server/statvar_hierarchy_util.go

```diff
@@ -0,0 +1,107 @@
+// Copyright 2021 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package server
+
+import (
+	"context"
+	"strings"
+
+	"cloud.google.com/go/bigtable"
+	pb "github.com/datacommonsorg/mixer/internal/proto"
+	"github.com/datacommonsorg/mixer/internal/util"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+	"google.golang.org/protobuf/encoding/protojson"
+)
+
+// RankingInfo holds the ranking information for each stat var hierarchy search
+// result.
+type RankingInfo struct {
+	ApproxNumPv int
+	RankingName string
+}
+
+// GetRawSvg gets the raw svg mapping.
+func GetRawSvg(ctx context.Context, baseTable *bigtable.Table) (map[string]*pb.StatVarGroupNode, error) {
+	svgResp := &pb.StatVarGroups{}
+	row, err := baseTable.ReadRow(ctx, util.BtStatVarGroup)
+	if err != nil {
+		return nil, err
+	}
+	if len(row[util.BtFamily]) == 0 {
+		return nil, status.Errorf(codes.NotFound, "Stat Var Group not found in cache")
+	}
+	raw := row[util.BtFamily][0].Value
+	jsonRaw, err := util.UnzipAndDecode(string(raw))
+	if err != nil {
+		return nil, err
+	}
+	err = protojson.Unmarshal(jsonRaw, svgResp)
+	if err != nil {
+		return nil, err
+	}
+	return svgResp.StatVarGroups, nil
+}
+
+// GetParentSvgMap gets the mapping of svg/sv id to the parent svg for that svg/sv.
+func GetParentSvgMap(rawSvg map[string]*pb.StatVarGroupNode) map[string]string {
+	parentSvgMap := map[string]string{}
+	for svgID, svgData := range rawSvg {
+		for _, childSvg := range svgData.ChildStatVarGroups {
+			if _, ok := parentSvgMap[childSvg.Id]; !ok {
+				parentSvgMap[childSvg.Id] = svgID
```

Comment:
So now only one parent is stored. What is the implication of the parent selection? It came to mind to store all the parents, but looks like for now, only a random random parent would be used?

Reply:
We were always only choosing a single path so I thought we would only need one parent stored? Would it be better to store all the parents?

Reply:
I think it's better, store a sorted parent and client pick the first one for now.

Reply:
done